File: backend/routes/phase8_api.py
"""
Phase 8 API Routes: End-to-End Testing & Production Readiness
"""

from fastapi import APIRouter, HTTPException
from typing import Dict, Any, List
from backend.production_readiness import (
    ProductionReadinessChecker,
    HealthMonitor,
    IntegrationValidator,
)

router = APIRouter(prefix="/api/phase8", tags=["Phase 8: E2E Testing & Production Readiness"])

readiness_checker = ProductionReadinessChecker()
health_monitor = HealthMonitor()
integration_validator = IntegrationValidator()


@router.get("/summary")
async def get_phase8_summary() -> Dict[str, Any]:
    """Get Phase 8 summary"""
    return {
        "phase": "Phase 8: End-to-End Testing & Production Readiness",
        "description": "Comprehensive testing and production readiness validation",
        "components": [
            "E2E Test Suite",
            "Production Readiness Checker",
            "Health Monitor",
            "Integration Validator",
        ],
        "status": "operational",
        "endpoints": 15,
    }


@router.get("/readiness/checks")
async def get_readiness_checks() -> Dict[str, Any]:
    """Get all production readiness checks"""
    return await readiness_checker.run_all_checks()


@router.get("/readiness/checklist")
async def get_readiness_checklist() -> List[Dict[str, Any]]:
    """Get production readiness checklist"""
    return readiness_checker.get_checklist()


@router.get("/readiness/summary")
async def get_readiness_summary() -> Dict[str, Any]:
    """Get production readiness summary"""
    return readiness_checker.get_summary()


@router.get("/health/system")
async def get_system_health() -> Dict[str, Any]:
    """Get overall system health"""
    return await health_monitor.get_system_health()


@router.get("/health/history")
async def get_health_history(limit: int = 10) -> List[Dict[str, Any]]:
    """Get health history"""
    return await health_monitor.get_health_history(limit)


@router.get("/health/component/{component}")
async def get_component_health(component: str) -> Dict[str, Any]:
    """Get health of a specific component"""
    return await health_monitor.get_component_health(component)


@router.get("/health/metrics")
async def get_health_metrics() -> Dict[str, Any]:
    """Get health metrics summary"""
    return await health_monitor.get_metrics_summary()


@router.get("/integrations/validate")
async def validate_all_integrations() -> Dict[str, Any]:
    """Validate all integrations"""
    return await integration_validator.validate_all_integrations()


@router.get("/integrations/map")
async def get_integration_map() -> Dict[str, Any]:
    """Get integration map"""
    return integration_validator.get_integration_map()


@router.get("/integrations/{integration_id}")
async def validate_integration(integration_id: str) -> Dict[str, Any]:
    """Validate a specific integration"""
    return await integration_validator.validate_integration(integration_id)


@router.get("/tests/status")
async def get_test_status() -> Dict[str, Any]:
    """Get E2E test status"""
    return {
        "test_suite": "Phase 8 E2E Tests",
        "total_tests": 50,
        "test_categories": [
            "Phase 0: Boot Stability",
            "Phase 1: Pillar Hardening",
            "Phase 2: RAG & Memory",
            "Phase 3: Learning Engine",
            "Phase 4: Copilot",
            "Phase 5: World Builder UI",
            "Phase 6: Enterprise API",
            "Phase 7: SaaS Readiness",
            "End-to-End User Flows",
            "Production Readiness",
        ],
        "status": "ready",
        "last_run": None,
    }


@router.get("/tests/results")
async def get_test_results() -> Dict[str, Any]:
    """Get E2E test results"""
    return {
        "test_suite": "Phase 8 E2E Tests",
        "status": "not_run",
        "message": "Run tests with: pytest tests/e2e/test_critical_flows.py -v",
        "total_tests": 50,
        "passed": 0,
        "failed": 0,
        "skipped": 0,
    }


@router.get("/status")
async def get_phase8_status() -> Dict[str, Any]:
    """Get overall Phase 8 status"""
    readiness = await readiness_checker.run_all_checks()
    health = await health_monitor.get_system_health()
    integrations = await integration_validator.validate_all_integrations()
    
    return {
        "phase": "Phase 8: End-to-End Testing & Production Readiness",
        "timestamp": health["timestamp"],
        "overall_status": "operational",
        "readiness": {
            "status": readiness["overall_status"],
            "score": readiness["readiness_score"],
            "total_checks": readiness["total_checks"],
            "passed_checks": readiness["passed_checks"],
        },
        "health": {
            "status": health["status"],
            "cpu_percent": health["components"]["system"]["cpu_percent"],
            "memory_percent": health["components"]["system"]["memory_percent"],
        },
        "integrations": {
            "status": integrations["overall_status"],
            "success_rate": integrations["success_rate"],
            "total_tests": integrations["total_tests"],
            "passed_tests": integrations["passed_tests"],
        },
        "e2e_tests": {
            "status": "ready",
            "total_tests": 50,
        }
    }

File: frontend/src/api/phase8Api.ts
import { apiUrl, WS_BASE_URL } from './config';
/**
 * Phase 8 API Client: End-to-End Testing & Production Readiness
 */

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || apiUrl('';

export interface Phase8Summary {
  phase: string;
  description: string;
  components: string[];
  status: string;
  endpoints: number;
}

export interface ReadinessCheck {
  id: string;
  name: string;
  category: string;
  status: string;
  checks: Array<{
    name: string;
    status: string;
    message: string;
  }>;
}

export interface ReadinessResult {
  timestamp: string;
  overall_status: string;
  total_checks: number;
  passed_checks: number;
  failed_checks: number;
  readiness_score: number;
  categories: Record<string, {
    total: number;
    passed: number;
    failed: number;
  }>;
  checks: ReadinessCheck[];
}

export interface SystemHealth {
  timestamp: string;
  status: string;
  components: Record<string, {
    status: string;
    message?: string;
    cpu_percent?: number;
    memory_percent?: number;
    disk_percent?: number;
  }>;
  metrics: {
    uptime_seconds: number;
    total_requests: number;
    error_rate: number;
    avg_response_time_ms: number;
  };
  warnings?: string[];
  unhealthy_components?: string[];
}

export interface IntegrationTest {
  id: string;
  name: string;
  description: string;
  phases: string[];
  critical: boolean;
  status?: string;
  message?: string;
  duration_ms?: number;
}

export interface IntegrationResult {
  timestamp: string;
  overall_status: string;
  total_tests: number;
  passed_tests: number;
  failed_tests: number;
  skipped_tests: number;
  success_rate: number;
  critical_failures: string[];
  tests: IntegrationTest[];
}

export interface Phase8Status {
  phase: string;
  timestamp: string;
  overall_status: string;
  readiness: {
    status: string;
    score: number;
    total_checks: number;
    passed_checks: number;
  };
  health: {
    status: string;
    cpu_percent: number;
    memory_percent: number;
  };
  integrations: {
    status: string;
    success_rate: number;
    total_tests: number;
    passed_tests: number;
  };
  e2e_tests: {
    status: string;
    total_tests: number;
  };
}

export class Phase8ApiClient {
  private baseUrl: string;

  constructor(baseUrl: string = API_BASE_URL) {
    this.baseUrl = baseUrl;
  }

  async getSummary(): Promise<Phase8Summary> {
    const response = await fetch(`${this.baseUrl}/api/phase8/summary`);
    if (!response.ok) throw new Error('Failed to fetch Phase 8 summary');
    return response.json();
  }

  async getReadinessChecks(): Promise<ReadinessResult> {
    const response = await fetch(`${this.baseUrl}/api/phase8/readiness/checks`);
    if (!response.ok) throw new Error('Failed to fetch readiness checks');
    return response.json();
  }

  async getReadinessChecklist(): Promise<ReadinessCheck[]> {
    const response = await fetch(`${this.baseUrl}/api/phase8/readiness/checklist`);
    if (!response.ok) throw new Error('Failed to fetch readiness checklist');
    return response.json();
  }

  async getReadinessSummary(): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/phase8/readiness/summary`);
    if (!response.ok) throw new Error('Failed to fetch readiness summary');
    return response.json();
  }

  async getSystemHealth(): Promise<SystemHealth> {
    const response = await fetch(`${this.baseUrl}/api/phase8/health/system`);
    if (!response.ok) throw new Error('Failed to fetch system health');
    return response.json();
  }

  async getHealthHistory(limit: number = 10): Promise<SystemHealth[]> {
    const response = await fetch(`${this.baseUrl}/api/phase8/health/history?limit=${limit}`);
    if (!response.ok) throw new Error('Failed to fetch health history');
    return response.json();
  }

  async getComponentHealth(component: string): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/phase8/health/component/${component}`);
    if (!response.ok) throw new Error('Failed to fetch component health');
    return response.json();
  }

  async getHealthMetrics(): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/phase8/health/metrics`);
    if (!response.ok) throw new Error('Failed to fetch health metrics');
    return response.json();
  }

  async validateAllIntegrations(): Promise<IntegrationResult> {
    const response = await fetch(`${this.baseUrl}/api/phase8/integrations/validate`);
    if (!response.ok) throw new Error('Failed to validate integrations');
    return response.json();
  }

  async getIntegrationMap(): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/phase8/integrations/map`);
    if (!response.ok) throw new Error('Failed to fetch integration map');
    return response.json();
  }

  async validateIntegration(integrationId: string): Promise<IntegrationTest> {
    const response = await fetch(`${this.baseUrl}/api/phase8/integrations/${integrationId}`);
    if (!response.ok) throw new Error('Failed to validate integration');
    return response.json();
  }

  async getTestStatus(): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/phase8/tests/status`);
    if (!response.ok) throw new Error('Failed to fetch test status');
    return response.json();
  }

  async getTestResults(): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/phase8/tests/results`);
    if (!response.ok) throw new Error('Failed to fetch test results');
    return response.json();
  }

  async getStatus(): Promise<Phase8Status> {
    const response = await fetch(`${this.baseUrl}/api/phase8/status`);
    if (!response.ok) throw new Error('Failed to fetch Phase 8 status');
    return response.json();
  }
}

export const phase8Api = new Phase8ApiClient();
